Improve opcode reference detection robustness.

pw4ever opened the pull request:
The current opcode regexp fails to detect opcode reference in
the following cases.
* Texts extracted using non-Poppler (Glyph & Cog) `pdftotext` in
  Windows, which sometimes omit the headnote and begin directly with
  the opcode itself.
* Instructions that have a different headnote pattern, e.g., VMX
  instructions.

This changeset updates the opcode regexp to address the
above issues.

pw4ever committed:
Improve opcode reference detection robustness.

The current opcode regexp fails to detect opcode reference in
the following cases.
* Texts extracted using non-Poppler (Glyph & Cog) `pdftotext` in
  Windows, which sometimes omit the headnote and begin directly with
  the opcode itself.
* Instructions that have a different headnote pattern, e.g., VMX
  instructions.

This changeset updates the opcode regexp to address the
above issues.

skeeto commented:
Poppler `pdftotext` takes priority over non-Poppler since 1) Poppler is the backend of the original `pdftotext` and 2) I personally use the Poppler version. I'd like to support both, but it's vital that the Poppler version works *at least* as well as it already does.

Unfortunately your change causes the Poppler database to lose 69 mnemonics (see [pdftotext.txt](https://github.com/skeeto/x86-lookup/files/1833853/pdftotext.txt)). It also changes the page for the ambiguous `movsd` mnemonic, though I don't mind that. I produced that diff by printing each database with `pp` and running `diff -u` on the results.

pw4ever commented:
> Poppler pdftotext takes priority over non-Poppler since 1) Poppler is the backend of the original pdftotext and 2) I personally use the Poppler version. I'd like to support both, but it's vital that the Poppler version works at least as well as it already does.

I would love to have an easy-to-deploy (e.g., via Chocolatey) Windows port of Poppler. For now, I am stuck with `xpdf-utils`, which is Glyph & Cog version of it.

> Unfortunately your change causes the Poppler database to lose 69 mnemonics (see pdftotext.txt). It also changes the page for the ambiguous movsd mnemonic, though I don't mind that. I produced that diff by printing each database with pp and running diff -u on the results.

Thanks for the list; it is helpful narrowing down the issue.

Interestingly, there are some instructions, such as `prefetchwtl`, `invept`, `vmcall`, that the current regexp misses while the patch's regexp captures.
![image](https://user-images.githubusercontent.com/3914554/37730880-fd373a18-2cfd-11e8-9347-9365b0a89dcf.png)

I will dig into this and get back to you when I get time.

pw4ever committed:
Improve opcode reference detection robustness.

The current opcode regexp fails to detect opcode reference in
the following cases.
* Texts extracted using non-Poppler (Glyph & Cog) `pdftotext` in
  Windows, which sometimes omit the headnote and begin directly with
  the opcode itself.
* Instructions that have a different headnote pattern, e.g., VMX
  instructions in Volume 3.

This changeset updates the opcode regexp to addresses the
above issues.

pw4ever commented:
I have root-caused the regression to long opcode summary (after --) got split into multiple lines by pdftotext.

I have updated the regexp accordingly to address this. Now it finds all previously missing opcodes, plus the VMX ones in Volume 3 that are missing in current implementation.

![image](https://user-images.githubusercontent.com/3914554/37733601-e001f32c-2d05-11e8-89f8-c3486f33df68.png)

P.S. Test Elisp code in *scratch*.

```elisp

(require 'cl)

(setq oldindex
      (with-current-buffer (find-file-noselect "e:\\old.txt")
        (goto-char (point-min))
        (read (current-buffer))))

(setq newindex
      (with-current-buffer (find-file-noselect "e:\\new.txt")
        (goto-char (point-min))
        (read (current-buffer))))


(setq oldindex-newindex
      (set-difference oldindex newindex
                      :key #'car :test #'string=))


(setq newindex-oldindex
      (set-difference newindex oldindex
                      :key #'car :test #'string=))
```

skeeto commented:
Merged, thanks! It still misses a few mnemonics in my two-year-old copy 
of the manual, but it works exactly as advertised on the latest version. 
I suspect the old manual had formatting irregularities that have since 
been corrected by Intel and so no longer need to be accounted for in the 
regexp.

pw4ever commented:
Thanks for the quick turnaround!

> I suspect the old manual had formatting irregularities 

The SDM is manually formatted using Word rather than *TeX; so irregularity is no surprise. It is better to have lax/forgiving regexp to capture the essence.
